### PR TITLE
Created controller and service layer for doctor(POST)

### PR DESCRIPTION
**Pull request by Shahana2601**
Created controller and service layer for creating doctor endpoint.

### CODE REVIEW

File: src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java
Comment (divya-sathyan) [inline src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java:1]: Rename Poc or remove it

Comment (Shahana2601) [inline src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java:1]: Its already given.

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -1 +1,3 @@
 spring.application.name=HealthCare-POC
+spring.data.mongodb.uri=mongodb://localhost:27017/trainingPoc
+spring.data.mongodb.uuid-representation=standard
```
Comment (divya-sathyan) [inline src/main/resources/application.properties]: Add an extra line

Comment (Shahana2601) [reply src/main/resources/application.properties]: Okay I added

File: src/main/java/com/theelixrlabs/healthcare/constants/DoctorConstants.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package com.theelixrlabs.healthcare.constants;
+
+/**
+ * Constant class containing error messages and patters related to doctor.
+ */
+public class DoctorConstants {
+    public static final String AADHAAR_ALREADY_PRESENT = "Aadhaar already present in database";
+    public static final String AADHAAR_NUMBER_PATTERN = "Aadhaar number must be only 12 digits!";
+    public static final String AADHAAR_SHOULD_BE_MANDATORY = "Aadhaar should be mandatory!";
+    public static final String CREATE_DOCTOR_END_POINT = "/doctors";
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/constants/DoctorConstants.java]: alternate line 10-11 to 11-10 to make the class in proper alphabetic order

File: src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java
Comment (ashutosh-2609) [reply src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java:1]: Make sure the change for main class name should be pushed either from any existing PRs

File: src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Rest Controller for handling HTTP request and response.
+ */
+@RestController
+public class DoctorController {
+
+    @Autowired
+    DoctorService doctorService;
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java]: please make it private final

File: src/main/java/com/theelixrlabs/healthcare/constants/DoctorConstants.java (hunk 2 above)
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/constants/DoctorConstants.java]: POST end point should be correct
make sure it defines the purpose of API

File: src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.UUID;
+
+/**
+ * Data transfer object representing doctor information.
+ */
+@Getter
+@Setter
+@NoArgsConstructor
+public class DoctorDto {
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java:18]: Please add one line spaces after each variable declaration

File: src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    private UUID id;
+    @NotBlank(message = DoctorConstants.FIRST_NAME_SHOULD_BE_MANDATORY)
+    @Pattern(regexp = "^[a-zA-Z]+$", message = DoctorConstants.ONLY_CHARACTER)
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java]: please keep these regex patterns in constants

File: src/main/java/com/theelixrlabs/healthcare/response/FailureResponse.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Data;
+import java.util.List;
+
+/**
+ * A failure response class to wrap failure API responses.
+ */
+@Data
+@AllArgsConstructor
+public class FailureResponse {
+   private Boolean success;
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/response/FailureResponse.java]: Make use of primitive boolean
so the default values can be used

File: src/main/java/com/theelixrlabs/healthcare/response/SuccessResponse.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Data;
+
+/**
+ * A generic success response class to wrap success API responses.
+ * @param <T> Type of data contained in response.
+ */
+@Data
+@AllArgsConstructor
+public class SuccessResponse<T> {
+    private Boolean success;
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/response/SuccessResponse.java]: make use of boolean

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+     * Saves a new doctor document based on the provided DoctorDto.
+     *
+     * @param doctorDto : DoctorDto object containing doctor information.
+     * @return The saved dto object.
+     */
+    public DoctorDto saveDoctor(DoctorDto doctorDto) throws RuntimeException {
+        if (doctorRepository.findByAadhaar(doctorDto.getAadhaar()).isPresent()) {
+            throw new RuntimeException(DoctorConstants.AADHAAR_ALREADY_PRESENT);
+        }
+        DoctorEntity doctorEntity = new DoctorEntity();
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: Make use of @SuperBuilder in DoctorEntity to build objects

File: src/main/java/com/theelixrlabs/healthcare/constants/DoctorConstants.java (hunk 2 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/constants/DoctorConstants.java]: Yes I changed.

File: src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java (hunk 4 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java:18]: Yes added line spaces.

File: src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java (hunk 5 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java]: Yes I added in DoctorConstants file.

File: src/main/java/com/theelixrlabs/healthcare/response/FailureResponse.java (hunk 6 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/response/FailureResponse.java]: changed to boolean.

File: src/main/java/com/theelixrlabs/healthcare/response/SuccessResponse.java (hunk 7 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/response/SuccessResponse.java]: changed to boolean.

File: src/main/java/com/theelixrlabs/healthcare/constants/DoctorConstants.java (hunk 2 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/constants/DoctorConstants.java]: Okay I changed to saveDoctor

File: src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java
Comment (TruptiPatil015) [inline src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java:1]: please use constructor based injection to autowire service.

File: src/main/java/com/theelixrlabs/healthcare/response/SuccessResponse.java
Comment (TruptiPatil015) [inline src/main/java/com/theelixrlabs/healthcare/response/SuccessResponse.java:1]: Usage of primitive datatype like boolean would be better as it saves memory.

File: src/main/java/com/theelixrlabs/healthcare/response/FailureResponse.java
Comment (TruptiPatil015) [inline src/main/java/com/theelixrlabs/healthcare/response/FailureResponse.java:1]: Here as well.

File: src/main/resources/application.properties
Comment (TruptiPatil015) [inline src/main/resources/application.properties:1]: Remove extra line from end.

File: src/main/resources/application.properties (hunk 1 above)
Comment (TruptiPatil015) [reply src/main/resources/application.properties]: please remove line.

File: src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java (hunk 3 above)
Comment (TruptiPatil015) [reply src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java]: It is recommended to use constructor based injection.

File: src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java
Comment (TruptiPatil015) [inline src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java:1]: Suggestion:
The handler for generic exceptions (Exception.class) might catch exceptions you don't intend to handle. It's better to be specific or have a more generic handler at a lower priority.

File: src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java [hunk 9]
```diff
@@ -1,7 +1,9 @@
 package com.theelixrlabs.healthcare;
 
+import org.modelmapper.ModelMapper;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
 
 @SpringBootApplication
```
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java]: Remove unwanted lines

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.stereotype.Service;
+
+/**
+ * Service class responsible for handling operations related to doctor.
+ */
+@Service
+public class DoctorService {
+    @Autowired
+    private DoctorRepository doctorRepository;
+
```
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: Use Constructor injection

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    /**
+     * Saves a new doctor document based on the provided DoctorDto.
+     *
+     * @param doctorDto : DoctorDto object containing doctor information.
+     * @return The saved dto object.
+     */
+    public DoctorDto saveDoctor(DoctorDto doctorDto) throws RuntimeException {
+        if (doctorRepository.findByAadhaar(doctorDto.getAadhaar()).isPresent()) {
+            throw new RuntimeException(DoctorConstants.AADHAAR_ALREADY_PRESENT);
+        }
```
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java:27]: Do not throw RuntiemExcpetion directly, try handling it, and returning the response with status code bad request

File: src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    @NotBlank(message = DoctorConstants.LAST_NAME_SHOULD_BE_MANDATORY)
+    @Pattern(regexp = "^[a-zA-Z]+$", message = DoctorConstants.ONLY_CHARACTER)
+    private String lastName;
+    @NotBlank(message = DoctorConstants.DEPARTMENT_SHOULD_BE_MANDATORY)
+    @Pattern(regexp = "^[a-zA-Z]+$", message = DoctorConstants.ONLY_CHARACTER)
+    private String department;
+    @NotNull(message = DoctorConstants.AADHAAR_SHOULD_BE_MANDATORY)
+    @Pattern(regexp = "^[2-9]{1}[0-9]{11}$", message = DoctorConstants.AADHAAR_NUMBER_PATTERN)
+    private String aadhaar;
+}
```
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java]: Rename the varibale aadhaar to aadhaarNumber

**Review by vinodkadam22 — CHANGES_REQUESTED** [review]
Resolve your PR conflicts

File: src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java (hunk 9 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/HealthCarePocApplication.java]: Removed unwanted line.

File: src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java (hunk 12 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java]: Changed to aadhaarNumber

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java (hunk 10 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: Okay. Added constructor based injection.

File: src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java (hunk 3 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/controller/DoctorController.java]: Yes I removed @Autowired and implemented constructor based injection.

File: src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java:1]: Yes I implemented custom exception handler.

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java (hunk 8 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: Implemented @SuperBuilder in DoctorEntity.

File: src/main/resources/application.properties (hunk 1 above)
Comment (Shahana2601) [reply src/main/resources/application.properties]: Removed the line.

File: src/main/java/com/theelixrlabs/healthcare/exceptionHandler/CustomException.java [hunk 13]
```diff
@@ -0,0 +1,10 @@
+package com.theelixrlabs.healthcare.exceptionHandler;
+
+/**
+ * Custom Exception class for handling exception.
+ */
+public class CustomException extends RuntimeException{
+    public CustomException(String ExceptionMessage) {
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/exceptionHandler/CustomException.java]: ExceptionMessage should be in camel case

exceptionMessage

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java [hunk 14]
```diff
@@ -1,37 +1,59 @@
...
-        doctorEntity.setLastName(doctorDto.getLastName());
-        doctorEntity.setDepartment(doctorDto.getDepartment());
-        doctorEntity.setAadhaar(doctorDto.getAadhaar());
+        UUID uuid = UUID.randomUUID();
+        DoctorEntity doctorEntity = DoctorEntity.builder().id(uuid).firstName(doctorDto.getFirstName()).lastName(doctorDto.getLastName()).department(doctorDto.getDepartment()).aadhaarNumber(doctorDto.getAadhaarNumber()).build();
         doctorRepository.save(doctorEntity);
-        doctorDto.setId(doctorEntity.getId());
-        return doctorDto;
+        DoctorDto savedDoctorDto;
+        savedDoctorDto = DoctorDto.builder().id(doctorEntity.getId()).firstName(doctorEntity.getFirstName()).lastName(doctorEntity.getLastName()).department(doctorEntity.getDepartment()).aadhaarNumber(doctorEntity.getAadhaarNumber()).build();
```
Comment (ashutosh-2609) [inline src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: line 35 and 38 split into multiple lines 
make it readable

File: src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.UUID;
+
+/**
+ * Data transfer object representing doctor information.
+ */
+@Getter
+@Setter
+@NoArgsConstructor
+@Data
+@SuperBuilder
```
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java]: Do you actually need @Getter and @Setter when you have add @Data ?

File: src/main/java/com/theelixrlabs/healthcare/exceptionHandler/CustomException.java (hunk 13 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/exceptionHandler/CustomException.java]: Yes I changed to exceptionMessage.

File: src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java:1]: Reformat your code and remove additional lines

**Review by vinodkadam22 — CHANGES_REQUESTED** [review]

**Comment by vinodkadam22** [comment]
And resolve your PR conflicts

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java (hunk 14 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: Okay. Changed

File: src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java (hunk 15 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/dto/DoctorDto.java]: No. there is no need of additional @getter @setter. I removed those.

File: src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java:1]: Reformatted the codes.

File: src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java [hunk 16]
```diff
@@ -37,9 +38,11 @@ public ResponseEntity<FailureResponse> handleValidationException(MethodArgumentN
...
-    @ExceptionHandler(Exception.class)
-    public ResponseEntity<FailureResponse> handleIllegalArgumentException(Exception exception) {
+    @ExceptionHandler(CustomException.class)
+    public ResponseEntity<FailureResponse> handleAadhaarAlreadyPresentException(CustomException exception) {
         List<String> errorMessages = Arrays.asList(exception.getMessage().split(DoctorConstants.COMMA_DELIMITER));
-        return ResponseEntity.badRequest().body(new FailureResponse(false, errorMessages));
+        return new ResponseEntity<>(new FailureResponse(false,errorMessages), HttpStatus.BAD_REQUEST);
     }
+
+
```
Comment (divya-sathyan) [inline src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java]: Remove extra lines

Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/exceptionHandler/GlobalExceptionHandler.java]: Removed extra lines.

**Review by ashutosh-2609 — APPROVED** [review]

**Review by TruptiPatil015 — APPROVED** [review]

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * Custom validation method for checking regex pattern of firstname lastname and Aadhaar number
+     *
+     * @param doctorDto : DoctorDto object containing doctor information.
+     * @throws CustomException : Class to handle custom exception
+     */
+    public void validateDoctor(DoctorDto doctorDto) throws CustomException {
+        if (!(doctorDto.getFirstName().matches(DoctorConstants.CHARACTER_ONLY_REGEX_PATTERN))) {
+            throw new CustomException(DoctorConstants.FIRST_NAME_MUST_BE_CHARACTER);
```
Comment (vinodkadam22) [inline src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: Make it private method

**Review by vinodkadam22 — CHANGES_REQUESTED** [review]

**Review by EXR-85 — APPROVED** [review]

File: src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java (hunk 17 above)
Comment (Shahana2601) [reply src/main/java/com/theelixrlabs/healthcare/service/DoctorService.java]: resolved

**Pull request merged**